File: PilhaDinamica/Java/PilhaDinamica.java
package src.PilhaDinamica.Java;

public class PilhaDinamica<T> {
    private T[] dados;
    private int ponteiroFim;

    public PilhaDinamica(){
        dados = (T[]) new Object[4];
        ponteiroFim = -1;
    }

    public boolean estaCheia(){
        return ponteiroFim == (dados.length-1);
    }

    public boolean estaVazia(){
        return ponteiroFim == -1;
    }

    private void redimencionar(){
        T[] novosDados = (T[]) new Object[dados.length*2];

        for (int i = 0; i <= ponteiroFim; i++) {
            novosDados[i] = dados[i];
        }

        dados = novosDados;
    }

    public void inserir(T dado){
        if (estaCheia()) redimencionar();
        ponteiroFim++;
        dados[ponteiroFim] = dado;
    }

    public T remover(){
        T retorno = null;

        if (!estaVazia()) {
            retorno = dados[ponteiroFim];
            dados[ponteiroFim] = null;
            ponteiroFim--;
        } else
            System.out.println("Pilha Vazia!");

        return retorno;
    }

    public T obterTopo(){
        T retorno = null;

        if (!estaVazia())
            retorno = dados[ponteiroFim];
        else
            System.out.println("Pilha Vazia!");

        return retorno;
    }

    public void imprimirPilha(){
        for (int i = ponteiroFim; i >= 0; i--) {
            System.out.println(dados[i]);
        }
    }

    public int tamanhoPilha(){
        return ponteiroFim+1;
    }
}

File: PilhaDinamica/Kotlin/PilhaDinamica.kt
class PilhaDinamica(var tamanho: Int) : Empilhavel {

    var ponteiroTopo: NoDuplo? = null
    var quantidade = 0

    override fun empilhar(dado: Any?) {
        if (!estaCheia()) {
            val novoNo = NoDuplo(dado)
            if (!estaVazia()) {
                ponteiroTopo?.proximo = novoNo
                novoNo.anterior = ponteiroTopo
            }
            ponteiroTopo = novoNo
            quantidade++
        } else {
            println("Pilha Cheia")
        }
    }

    override fun desempilhar(): Any? {
        var aux: Any? = null

        if (!estaVazia()) {
            aux = ponteiroTopo?.dado
            ponteiroTopo = ponteiroTopo?.anterior
            if (ponteiroTopo != null)
                ponteiroTopo?.proximo = null
            quantidade--
        } else {
            println("Pilha Vazia")
        }

        return aux
    }

    override fun espiar(): Any? {
        return if (!estaVazia()) {
            ponteiroTopo?.dado
        } else {
            println("Pilha Vazia")
            null
        }
    }

    override fun atualizar(dado: Any?) {
        if (!estaVazia()) {
            ponteiroTopo?.dado = dado
        } else {
            println("Pilha Vazia")
        }
    }

    override fun estaCheia(): Boolean {
        return quantidade == tamanho
    }

    override fun estaVazia(): Boolean {
        return quantidade == 0
    }

    override fun imprimir(): String {
        var aux = ponteiroTopo
        var retorno = "["

        while (aux != null) {
            retorno += aux.dado
            if (aux.anterior != null) retorno += ", "
            aux = aux.anterior
        }

        return "$retorno]"
    }
}